### ISSUE TITLE
Write permission gets lost during cleanup operation

### ISSUE THREAD

**Issue by jojoob** [opencast/opencast#3510]
**Describe the bug**
When creating an event via admin UI, uploading a video for the presentation track and select the fast workflow the video processes just fine.
On file system level however the permissions looking odd:

```bash
root@ocadmin:~# ls -l /opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/*
/opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/0:
total 12
-rw-r--r-- 2 opencast opencast  521 Mar  3 15:14 9e6e7e6e-832c-4d7d-ac98-a8417e71e8d4.xml
-rw-r--r-- 1 opencast opencast  830 Mar  3 15:14 manifest.xml
-rw-r--r-- 4 opencast opencast 2281 Mar  3 15:14 security-policy-episode.xml

/opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/1:
total 1928
-rw-r--r-- 2 opencast opencast     521 Mar  3 15:14 9e6e7e6e-832c-4d7d-ac98-a8417e71e8d4.xml
-r--r--r-- 1 opencast opencast 1956169 Mar  3 15:14 dc0dee77-121f-493b-94f4-627e65061e58.webm
-rw-r--r-- 1 opencast opencast    5160 Mar  3 15:16 manifest.xml
-rw-r--r-- 4 opencast opencast    2281 Mar  3 15:14 security-policy-episode.xml

root@ocadmin:~# find /opencast1/ ! -perm -u=w -path "*d9766e49-6b42-43d1-a253-641c6ea7908d*" -printf "%M %p\n"
-r--r--r-- /opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/1/dc0dee77-121f-493b-94f4-627e65061e58.webm
```

When starting a "Publish" workflow on this just uploaded and successful processed event the workflow fails.

<details>

```
job-operation
Distribute
job-type
org.opencastproject.distribution.download
stack-trace
org.opencastproject.serviceregistry.api.ServiceRegistryException: Error handling operation 'Distribute'
	at org.opencastproject.distribution.download.DownloadDistributionServiceImpl.process(DownloadDistributionServiceImpl.java:695)
	at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:313)
	at org.opencastproject.job.api.AbstractJobProducer$JobRunner.call(AbstractJobProducer.java:272)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.opencastproject.distribution.api.DistributionException: Unable to find http://ocadmin.rz.htw-berlin.de:8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/9e6e7e6e-832c-4d7d-ac98-a8417e71e8d4/dublincore.xml in the workspace
	at org.opencastproject.distribution.download.DownloadDistributionServiceImpl.distributeElement(DownloadDistributionServiceImpl.java:306)
	at org.opencastproject.distribution.download.DownloadDistributionServiceImpl.distributeElements(DownloadDistributionServiceImpl.java:268)
	at org.opencastproject.distribution.download.DownloadDistributionServiceImpl.process(DownloadDistributionServiceImpl.java:679)
	... 6 more
Caused by: org.opencastproject.util.NotFoundException: Error acquiring lock for /opencast1/workspace/http_ocadmin.rz.htw-berlin.de_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/9e6e7e6e-832c-4d7d-ac98-a8417e71e8d4/dublincore.xml
	at org.opencastproject.util.IoSupport.acquireLock(IoSupport.java:501)
	at org.opencastproject.util.IoSupport.locked(IoSupport.java:476)
	at org.opencastproject.workspace.impl.WorkspaceImpl.get(WorkspaceImpl.java:383)
	at org.opencastproject.workspace.impl.WorkspaceImpl.get(WorkspaceImpl.java:336)
	at org.opencastproject.distribution.download.DownloadDistributionServiceImpl.distributeElement(DownloadDistributionServiceImpl.java:304)
	... 8 more
Caused by: java.io.FileNotFoundException: /opencast1/workspace/http_ocadmin.rz.htw-berlin.de_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/9e6e7e6e-832c-4d7d-ac98-a8417e71e8d4/dublincore.xml (Permission denied)
	at java.base/java.io.RandomAccessFile.open0(Native Method)
	at java.base/java.io.RandomAccessFile.open(RandomAccessFile.java:345)
	at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:259)
	at java.base/java.io.RandomAccessFile.<init>(RandomAccessFile.java:214)
	at org.opencastproject.util.IoSupport.acquireLock(IoSupport.java:497)
	... 12 more
```

</details>

After the failed publish workflow the permissions look like:

```bash
root@ocadmin:~# find /opencast1/ ! -perm -u=w -path "*d9766e49-6b42-43d1-a253-641c6ea7908d*" -printf "%M %p\n"
-r--r--r-- /opencast1/workspace/http_ocadmin.example.com_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/94e2040b-06e4-4cab-b0bb-38c2bac1f69d/2022_03_03_15_11___Johannes__Display__OC_Studio__0_000s_player.jpg
-r--r--r-- /opencast1/workspace/http_ocadmin.example.com_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/c6c13737-f271-4a05-b9ab-a34067b94deb/2022_03_03_15_11___Johannes__Display__OC_Studio_.mp4
-r--r--r-- /opencast1/workspace/http_ocadmin.example.com_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/9e6e7e6e-832c-4d7d-ac98-a8417e71e8d4/dublincore.xml
-r--r--r-- /opencast1/workspace/http_ocadmin.example.com_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/security-policy-episode/xacml.xml
-r--r--r-- /opencast1/workspace/http_ocadmin.example.com_8080/files/mediapackage/d9766e49-6b42-43d1-a253-641c6ea7908d/24e03bb7-7ed3-4182-b26f-9c1ee18aa182/2022_03_03_15_11___Johannes__Display__OC_Studio__0_000s_search.jpg
-r--r--r-- /opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/2/dc0dee77-121f-493b-94f4-627e65061e58.webm
-r--r--r-- /opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/2/30b84f59-d0ad-4bc6-97f0-916df022f37a.webm
-r--r--r-- /opencast1/archive/mh_default_org/d9766e49-6b42-43d1-a253-641c6ea7908d/1/dc0dee77-121f-493b-94f4-627e65061e58.webm
-r--r--r-- /opencast1/downloads/mh_default_org/internal/d9766e49-6b42-43d1-a253-641c6ea7908d/c6c13737-f271-4a05-b9ab-a34067b94deb/2022_03_03_15_11___Johannes__Display__OC_Studio_.mp4
```

For troubleshooting purposes I modified the fast workflow step by step and explored that the "cleanup" operation is the one after which the write permission gets lost. When removing the cleanup operation from the fast workflow the webm file keeps its write permission. But this does not prevent a subsequent publish workflow from failing.

**Environment Information**
 - OS: Ubuntu 20.04
 - Browser Chromium
 - Opencast 11.4.0 - 248a5fc - built on 2022-03-03
 - Distributed setup across three servers (admin, presentation, worker) and one reverse proxy in front
 - No workflow modifications made
 - Ubuntu NFS server with "default" configuration

**Comment by jojoob** [comment]
I'm doing some tests with the help of Vagrant...

With the Vagrantfile below I installed NFS server and client on the same machine as well as Opencast allinone.

On filesystem level permissions look the same as with my problematic initial setup but I get no errors on subsequent workflows. Seems that the read-only permissions on some files does not necessarily lead to workflow failures.

<details>

<summary>
Vagrantfile
</summary>

```ruby
$script = <<-SCRIPT
apt-get update
apt-get install -y apt-transport-https ca-certificates sudo wget gnupg2
apt-get install -y nfs-kernel-server nfs-common
mkdir /srv/nfs_opencast
echo "/srv/nfs_opencast 127.0.0.1(rw,sync,no_subtree_check)" >> /etc/exports
exportfs -a
mkdir /srv/opencast
echo "127.0.0.1:/srv/nfs_opencast /srv/opencast nfs _netdev 0 0" >> /etc/fstab
mount -a
echo "deb https://pkg.opencast.org/debian 11.x stable" | sudo tee /etc/apt/sources.list.d/opencast.list
wget -qO - https://pkg.opencast.org/gpgkeys/opencast-deb.key | sudo apt-key add -
apt-get update
apt-get install -y opencast-11-allinone
cp /usr/share/opencast/docs/scripts/activemq/activemq.xml /etc/activemq/activemq.xml
chown opencast:opencast /srv/nfs_opencast
echo "org.ops4j.pax.web.listening.addresses=0.0.0.0" >> /etc/opencast/org.ops4j.pax.web.cfg
systemctl enable activemq.service
systemctl enable elasticsearch.service
systemctl enable opencast.service
systemctl start activemq.service
systemctl start elasticsearch.service
sleep 10
systemctl start opencast.service
SCRIPT

Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/focal64"
  config.vm.network "forwarded_port", guest: 8080, host: 8080
  config.vm.provider "virtualbox" do |vb|
    vb.memory = "4096"
  end

  config.vm.provision "shell", inline: $script
end
```

</details>

**Comment by jojoob** [comment]
With another Vagrant setup I figured out that whenever Ubuntu is used for NFS server **or** Opencast the permissions are mangled. Only when both servers are Debian the permissions look normal. ~~This leads me to the assumption that something in Ubuntus NFS packages causes this.~~ PS: I downloaded and installed the nfs-common debian package and its dependencies on an Ubuntu machine but observed the same problems.

<details>

<summary>
Vagrantfile
</summary>

```ruby
$provision_nfs = <<-SCRIPT
apt-get update
apt-get install -y nfs-kernel-server
mkdir /srv/nfs_opencast
chown 7967:7967 /srv/nfs_opencast
echo "/srv/nfs_opencast 192.168.56.4(rw,sync,no_subtree_check)" >> /etc/exports
exportfs -a
SCRIPT

$provision_opencast = <<-SCRIPT
apt-get update
apt-get install -y nfs-common
mkdir /srv/opencast
echo "192.168.56.9:/srv/nfs_opencast /srv/opencast nfs _netdev 0 0" >> /etc/fstab
mount -a
apt-get install -y apt-transport-https ca-certificates sudo wget gnupg2
echo "deb https://pkg.opencast.org/debian 11.x stable" | sudo tee /etc/apt/sources.list.d/opencast.list
wget -qO - https://pkg.opencast.org/gpgkeys/opencast-deb.key | sudo apt-key add -
apt-get update
apt-get install -y opencast-11-allinone
cp /usr/share/opencast/docs/scripts/activemq/activemq.xml /etc/activemq/activemq.xml
echo "org.ops4j.pax.web.listening.addresses=0.0.0.0" >> /etc/opencast/org.ops4j.pax.web.cfg
systemctl enable activemq.service
systemctl enable elasticsearch.service
systemctl enable opencast.service
systemctl start activemq.service
systemctl start elasticsearch.service
sleep 10
systemctl start opencast.service
SCRIPT

Vagrant.configure("2") do |config|
  config.vm.define "nfs" do |nfs|
    # nfs.vm.box = "ubuntu/focal64" # opencast: ubuntu -> mangled, opencast: debian -> mangled
    nfs.vm.box = "debian/bullseye64" # opencast: ubuntu -> mangled, opencast: debian -> ok
    nfs.vm.network "private_network", ip: "192.168.56.9"
    nfs.vm.provision "shell", inline: $provision_nfs
  end

  config.vm.define "opencast" do |opencast|
    opencast.vm.box = "ubuntu/focal64" # nfs: ubuntu -> mangled, nfs: debian -> mangled
    # opencast.vm.box = "debian/bullseye64" # nfs: ubuntu -> mangled, nfs: debian -> ok
    opencast.vm.network "private_network", ip: "192.168.56.4"
    opencast.vm.network "forwarded_port", guest: 8080, host: 8080
    opencast.vm.provider "virtualbox" do |vb|
      vb.memory = "4096"
    end
    opencast.vm.provision "shell", inline: $provision_opencast
  end
end
```

</details>

Mount options looking the same on a good and bad setup btw.

**Comment by jojoob** [comment]
Opencast 11.2 works fine while >=11.3 does not. Does not matter if installed from package or using Docker images.

I'm trying to find the commit that introduced the behaviour with git bisect... as soon as I found a time efficient way to build and test.

**Comment by jojoob** [comment]
#3328 introduced the bug. It has something to do with the `commons-io` dependency. Works with 2.8 but not with 2.10. Couldn't test 2.9 because of a dependency error (probably a bug within the commons-io package).

**Comment by jojoob** [comment]
After some git bisect sessions and poking in commons-io issue tracker I found the cause of this issue. It is already known and was tracked as [IO-751](https://issues.apache.org/jira/browse/IO-751). There is already a fix on master branch and it will be included in upcoming commons-io version 2.12. **Commons-io versions 2.9 to 2.11 are affected by the bug.**

As soon as commons-io 2.12 is released it should be save to update. I tested the fix using the current snapshot of commons-io.

**Background**
https://github.com/apache/commons-io/commit/62925e3084c9e72ae4659c00364a490252368326 fixed a bug in `PathUtils.setReadOnly()` method. The bug caused the method to set POSIX files to read only even it was called to do the opposite. Combined with the default delete option `OVERRIDE_READ_ONLY` (https://github.com/apache/commons-io/commit/8ae947bcc796d73b0eab7d94f10a002f66896af3) files were set to read only before deletion. But to make the chaos complete this did not happen due to another bug which was fixed with https://github.com/apache/commons-io/commit/580a7b5d395c315e4b7172f5563d3fe7461f92a6.
